nvflare/fuel/f3/cellnet/cell.py

```diff
@@ -232,26 +237,43 @@ def _get_result(self, req_id):
         waiter = self.requests_dict.pop(req_id)
         return waiter.result
 
-    def _future_wait(self, future, timeout):
+    def _check_error(self, future):
+        if future.error:
+            # must return a negative number
+            return -1
+        else:
+            return WaiterRC.OK
+
+    def _future_wait(self, future, timeout, abort_signal: Signal):
         # future could have an error!
         last_progress = 0
-        while not future.waiter.wait(timeout):
-            if future.error:
-                return False
-            current_progress = future.get_progress()
-            if last_progress == current_progress:
-                return False
+        while True:
+            rc = conditional_wait(future.waiter, timeout, abort_signal, condition_cb=self._check_error, future=future)
+            if rc == WaiterRC.IS_SET:
+                # waiter has been set!
+                break
+            elif rc == WaiterRC.TIMEOUT:
+                # timed out: check whether any progress has been made during this time
+                current_progress = future.get_progress()
+                if last_progress == current_progress:
```

Comment:
This should use another timeout value. On a slow network, it's likely that there is no progress for a long time.

Reply:
Not sure. In this case, the timeout value simply needs to be set to a big number.
It's true that different steps of the processing could use a different timeout value, but that would make the API very hard to use. We could just use a timeout value that accommodates the slowest step.